### PR TITLE
Adding metrics and feature flag for Log Persistence

### PR DESCRIPTION
**Pull request by Aayyush**
This PR adds the following metrics: 

- Storage Backend Read failures
- Storage Backend Write failures

In addition, we refactor the `LayeredJobStore` into `MemoryJobStore` which deals with jobs in memory and `StorageBackendJobStore` which deals with jobs in the storage backend. Lastly, we create a `FeatureAwareStorageBackend` that wraps the original storage backend implementation with a feature flag.

### CODE REVIEW

File: server/jobs/project_command_output_handler.go [hunk 1]
```diff
@@ -177,8 +180,10 @@ func (p *AsyncProjectCommandOutputHandler) GetReceiverBufferForPull(jobID string
 }
 
 func (p *AsyncProjectCommandOutputHandler) GetJob(jobID string) Job {
-	job, _ := p.jobStore.Get(jobID)
-	return job
+	if job, _ := p.jobStore.Get(jobID); job != nil {
+		return *job
```
Comment (nishkrishnan) [inline server/jobs/project_command_output_handler.go:184]: Why is job store returning a pointer if you're going to dereference it over here?

File: server/jobs/storage_backend.go [hunk 2]
```diff
@@ -19,7 +22,41 @@ type StorageBackend interface {
 	Read(key string) ([]string, error)
 
 	// Write logs to the storage backend
-	Write(key string, logs []string) (bool, error)
+	Write(key string, logs []string, fullRepoName string) (bool, error)
```
Comment (nishkrishnan) [inline server/jobs/storage_backend.go:25]: Can you make a ticket to remove the feature flag and therefore also this from the interface?

**Review by nishkrishnan — APPROVED** [review]

File: server/jobs/storage_backend.go (hunk 2 above)
Comment (Aayyush) [reply server/jobs/storage_backend.go:25]: Done!

File: server/jobs/project_command_output_handler.go (hunk 1 above)
Comment (msarvar) [reply server/jobs/project_command_output_handler.go:185]: Where is `GetJob` is being used? I searched in the codebase and seems it is only used by tests.

Comment (Aayyush) [reply server/jobs/project_command_output_handler.go:185]: It is used when we register new connections in the projectCommandOutputHandler() 
https://github.com/lyft/atlantis/blob/fa0e909c8ef8928d8dc68c369c8096be50ace354/server/jobs/project_command_output_handler.go#L131

Comment (msarvar) [reply server/jobs/project_command_output_handler.go:185]: That is `JobStore.Get(jobID)`, not `AsyncProjectCommandOutputHandler.GetJob`.

Comment (Aayyush) [reply server/jobs/project_command_output_handler.go]: Oh I got confused! yeah it's only being used for testing purposes. I have removed that method and updated the tests!

**Review by msarvar — APPROVED** [review]

**Pull request merged**